### PR TITLE
DTIFIT, BAMOS and DONSURF

### PR DESCRIPTION
**Pull request by xgrg**
DTIFIT:

-  `download_maps`: downloads FA, MD, RD and L1 maps in a structured folder (similar to `download_rc` in SPM)

BAMOS: 

- `volume()`
- `n_lesions()`
- `stats()`: extracts both lesion number and volume per region and depth layer
- `bullseye_plot()`

Thank you in advance @jhuguetn for your help reviewing this PR.

### CODE REVIEW

**Comment by xgrg** [comment]
Last commit includes `donsurf.py`: users can collect FreeSurfer-style surface-based measurements based on DONSURF cortical maps.

File: pyxnat/core/derivatives/donsurf.py [hunk 1]
```diff
@@ -0,0 +1,8 @@
+XNAT_RESOURCE_NAME = 'DONSURF'
+
+
+def aparc(self):
+    """Returns cortical features as estimated by `DONSURF`."""
+
+    from .freesurfer import aparc
+    return aparc(self)
```
Comment (jhuguetn) [inline pyxnat/core/derivatives/donsurf.py:8]: That's a bit mind-blowing, consider the following:
```suggestion
    from .freesurfer import fs_aparc
    return fs_aparc(self)
```

File: pyxnat/core/derivatives/dtifit.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+def download_maps(self, path):
+    '''Downloads a local copy of DWI maps (MD, FA, L1 ~axial, RD)'''
+    import os.path as op
+    rc_files = ['MD', 'FA', 'L1', 'RD']
+    for each in rc_files:
+        res = self._intf.array.experiments(experiment_id=self.parent().id(),
+                                           columns=['subject_label']).data[0]
+        subject_label = res['subject_label']
+        id = res['ID']
+        fp = '%s_%s_%s.nii.gz' % (subject_label, id, each)
```
Comment (jhuguetn) [inline pyxnat/core/derivatives/dtifit.py:13]: Shadows `id` builtin function, consider simply:
```suggestion
        fp = '%s_%s_%s.nii.gz' % (subject_label, res['ID'], each)
```

File: pyxnat/core/derivatives/freesurfer.py [hunk 3]
```diff
@@ -89,24 +89,44 @@ def aparc(self, atlas='desikan-killiany'):
...
         res2 = [e for e in res if e.startswith('# Measure ')]
         for each in volumes:
-            m = float([e for e in res2
-                       if each in e.split(', ')[1]][0].split(', ')[-2])
-            table.append([None, 'Volume_mm3', each, m])
+            m = [e for e in res2 if each == e.split(', ')[1]]
+            if len(m) > 1:
+                msg = 'Key %s is ambiguous in file %s (found entries: %s)'\
+                      % (each, f._uri, m)
+                raise Exception(msg)
```
Comment (jhuguetn) [inline pyxnat/core/derivatives/freesurfer.py:96]: Could we better use a more specific Exception type such as `ValueError`?

File: pyxnat/tests/test_resource_functions.py [hunk 4]
```diff
@@ -62,7 +62,21 @@ def test_fdg_quantification():
...
     r = e1.resource('BAMOS')
     v = r.volume()
-    assert(v == 33599.94964499201)
+    assert(v == 33620.32498628937)
+
+
+def test_bamos_stats():
+    r = e1.resource('BAMOS')
+    v = r.stats()
+    print(sum(v['volume']))
```
Comment (jhuguetn) [inline pyxnat/tests/test_resource_functions.py:71]:
```suggestion
```

File: pyxnat/tests/test_resource_functions.py [hunk 5]
```diff
@@ -62,7 +62,21 @@ def test_fdg_quantification():
...
+    assert(sum(v['volume']) == 33620.32429030311)
+    r.bullseye_plot(stats=v)
```
Comment (jhuguetn) [inline pyxnat/tests/test_resource_functions.py:73]: Shall we add a dummy check to assert it that worked? E.g.,
```
fig,ax = r.bullseye_plot(stats=v)
assert(isinstance(a,matplotlib.figure.Figure))
```

File: pyxnat/core/derivatives/bamos.py [hunk 6]
```diff
@@ -14,7 +28,201 @@ def volume(self):
...
+        return data
+
+    if stats is None:
+        stats = self.stats()
+
+    layers = [int(e) for e in set(stats['depth'])][1:]  # remove 0 (background)
+
+    # Highlight regions > nth percentile
+    if isinstance(segBold, int):
+        labels = [3, 1, 2, 4, 8, 6, 10, 9, 5, 7]
```
[the remark this answers is not in the document]
Comment (xgrg) [reply pyxnat/core/derivatives/bamos.py]: e339be9 should fix it.

File: pyxnat/core/derivatives/bamos.py [hunk 7]
```diff
@@ -14,7 +28,201 @@ def volume(self):
...
+        return label_values
+
+    def _ravel_stats_(stats):
+        ordered_labels = [2, 4, 8, 6, 10, 9, 5, 7, 3, 1]
+
+        data = []
+        for label in ordered_labels:
+            for depth in range(6):
+                data.append(stats[(depth, BAMOS_LABELS[label])])
+        return data
```
[the remark this answers is not in the document]
Comment (xgrg) [reply pyxnat/core/derivatives/bamos.py]: e339be9 should fix it.

File: pyxnat/core/derivatives/bamos.py [hunk 8]
```diff
@@ -14,7 +28,201 @@ def volume(self):
...
+        les = np.array(d.dataobj)
+
+        f = list(self.files('Layers_*.nii.gz'))[0]
+        f.get(fp)
+        d1 = np.array(nib.load(fp).dataobj)
+
+        f = list(self.files('Lobes_*.nii.gz'))[0]
+        f.get(fp)
+        d2 = np.array(nib.load(fp).dataobj)
+        os.unlink(fp)
```
[the remark this answers is not in the document]
Comment (xgrg) [reply pyxnat/core/derivatives/bamos.py]: 90dd3bb should do the job.

File: pyxnat/tests/test_resource_functions.py (hunk 4 above)
Comment (xgrg) [reply pyxnat/tests/test_resource_functions.py]: 939f706 should fix this.

File: pyxnat/tests/test_resource_functions.py (hunk 5 above)
Comment (xgrg) [reply pyxnat/tests/test_resource_functions.py]: 939f706 should fix this.

File: pyxnat/core/derivatives/dtifit.py (hunk 2 above)
Comment (xgrg) [reply pyxnat/core/derivatives/dtifit.py]: 939f706 fixes this.

File: pyxnat/core/derivatives/donsurf.py (hunk 1 above)
Comment (xgrg) [reply pyxnat/core/derivatives/donsurf.py]: 9d6cec1 fixes this.

File: pyxnat/core/derivatives/freesurfer.py (hunk 3 above)
Comment (xgrg) [reply pyxnat/core/derivatives/freesurfer.py]: Fine. Should be all in in dfeb78b.

**Comment by jhuguetn** [comment]
All good now from my side, if CI passes we can merge. Good job @xgrg

**Comment by xgrg** [comment]
Thank you @jhuguetn.

**Pull request merged**